service/src/main/java/org/cbioportal/service/impl/GenericAssayServiceImpl.java

```diff
@@ -90,20 +91,26 @@
 
         List<String> distinctMolecularProfileIds = molecularProfileIds.stream().distinct().sorted().collect(Collectors.toList());
 
-        List<String> commaSeparatedSampleIdsOfMolecularProfiles = molecularDataRepository
-            .getCommaSeparatedSampleIdsOfMolecularProfiles(distinctMolecularProfileIds);
-    
+        List<MolecularProfileSamples> commaSeparatedSampleIdsOfMolecularProfiles = molecularDataRepository
+                .getCommaSeparatedSampleIdsOfMolecularProfiles(distinctMolecularProfileIds);
+
+        Map<String, MolecularProfileSamples> commaSeparatedSampleIdsOfMolecularProfilesMap = commaSeparatedSampleIdsOfMolecularProfiles
+                .stream()
+                .collect(Collectors.toMap(MolecularProfileSamples::getMolecularProfileId, Function.identity()));
+
         Map<String, Map<Integer, Integer>> internalSampleIdsMap = new HashMap<>();
         List<Integer> allInternalSampleIds = new ArrayList<>();
-        
+
         for (int i = 0; i < distinctMolecularProfileIds.size(); i++) {
-            List<Integer> internalSampleIds = Arrays.stream(commaSeparatedSampleIdsOfMolecularProfiles.get(i).split(","))
-                .mapToInt(Integer::parseInt).boxed().collect(Collectors.toList());
+            String molecularProfileId = distinctMolecularProfileIds.get(i);
+            List<Integer> internalSampleIds = Arrays
+                    .stream(commaSeparatedSampleIdsOfMolecularProfilesMap.get(molecularProfileId).getSplitSampleIds())
+                    .mapToInt(Integer::parseInt).boxed().collect(Collectors.toList());
             HashMap<Integer, Integer> molecularProfileSampleMap = new HashMap<Integer, Integer>();
             for (int lc = 0; lc < internalSampleIds.size(); lc++) {
                 molecularProfileSampleMap.put(internalSampleIds.get(lc), lc);
             }
-            internalSampleIdsMap.put(distinctMolecularProfileIds.get(i), molecularProfileSampleMap);
+            internalSampleIdsMap.put(molecularProfileId, molecularProfileSampleMap);
```

Comment:
here is where we were relying on order yes?

Reply:
no at line 107